### PR TITLE
[WIP] Iam roles

### PR DESCRIPTION
**Pull request by RitchieVincent**

### CODE REVIEW

**Review by yadvr — CHANGES_REQUESTED** [review]
Following changes are advised @RitchieVincent:
- Add license header to all new files
- Change tab name to 'rules' or 'Rules'
- The DraggableRow component seems incomplete, should it cause an update to happen
- Avoid splitting the tab into so many components that it becomes unmanageable to follow
- Rename RoleAutoComplete to APIAutoComplete as essentially this components lists APIs. However there is no need to call the listApis (you can cache, definitely not everytime someone hits the add button), instead you may use `$store.getters.apis` since the APIs are discovered on login
- Rename RoleDelete to RuleDelete (essentially you're not deleting a role but a rule of that role)
- Refreshing the role does not cause the tab to refresh (perhaps add a watcher on the resource, call fetchData or equivalent to refresh the tab; reuse the fetchData in mounted)
- Remove the rule button but instead you can show an input component that takes in the input like the old UI, alternatively see the DetailSettingsTab to see how to add a wide/dotted add button that changes into an input component.
- In the list of available rules use the items from the listRolePermissions response not the listApis response.
- You may also consider if table is the right component, instead a list could be used to render the elements (for example in https://github.com/apache/cloudstack-primate/pull/38). My only issue with table is that it shrinks unevenly for smaller screens (the last columns get cut/hidden), if this can be solved with css usage of table is fine, as it also gives us pagination.
- Use loader for the row of component (or some feedback) while adding/updating/deleting a rule
- Use round button (shape=circle) for actions

**Comment by RitchieVincent** [comment]
@rhtyd I've updated the PR with the below changes:

- Add license header to all new files
**Complete**

- Change tab name to 'rules' or 'Rules'
**Complete**

- The DraggableRow component seems incomplete, should it cause an update to happen
**The dragging functionality now works, but I'm having an issue with the API, so I may need some assistance with that**

- Rename RoleAutoComplete to APIAutoComplete as essentially this components lists APIs. However there is no need to call the listApis (you can cache, definitely not everytime someone hits the add button), instead you may use $store.getters.apis since the APIs are discovered on login
**Complete**

- Rename RoleDelete to RuleDelete (essentially you're not deleting a role but a rule of that role)
**Complete**

- Refreshing the role does not cause the tab to refresh (perhaps add a watcher on the resource, call fetchData or equivalent to refresh the tab; reuse the fetchData in mounted)
**Complete**

- Remove the rule button but instead you can show an input component that takes in the input like the old UI, alternatively see the DetailSettingsTab to see how to add a wide/dotted add button that changes into an input component.
**Complete**

- You may also consider if table is the right component, instead a list could be used to render the elements (for example in apache#38). My only issue with table is that it shrinks unevenly for smaller screens (the last columns get cut/hidden), if this can be solved with css usage of table is fine, as it also gives us pagination.
**I've swapped this out for a list component now**

- Use loader for the row of component (or some feedback) while adding/updating/deleting a rule
**Complete**

- Use round button (shape=circle) for actions
**Complete**

**Comment by yadvr** [comment]
Thanks, I'll review and test the PR later today.

**Comment by yadvr** [comment]
@RitchieVincent there are many issues with the PR, will discuss over a call.

**Comment by RitchieVincent** [comment]
@rhtyd I've pushed up the latest action points that you've raised. That should hopefully be everything resolved now!

File: src/views/iam/RolePermissionTab.vue [hunk 1]
```diff
@@ -16,37 +16,404 @@
...
+      class="rules-list ant-list ant-list-bordered"
+      :class="{'rules-list--overflow-hidden' : updateTable}" >
+
+      <div class="rules-table-item ant-list-item">
+        <div class="rules-table__col rules-table__col--grab"></div>
+        <div class="rules-table__col rules-table__col--rule rules-table__col--new">
+          <a-p-i-auto-complete
+            @selectedRecord="onRuleSelect"
+            :error="newRuleSelectError"
+            :data="$store.getters.apis"
```
Comment (yadvr) [inline src/views/iam/RolePermissionTab.vue]: @RitchieVincent the api dropdown is very slow, can you instead create a variable in `data` and in `mounted` that the variable (i.e. api array data source) using `Object.keys($store.getters.apis)`?

Comment (yadvr) [reply src/views/iam/RolePermissionTab.vue]: Also if we can sort the API names list in ascending order.

**Comment by yadvr** [comment]
@RitchieVincent well done, looks good. Few minor issues:
- When update API is called, show the loading/spinner (either on editing the deny/accept dropdown for a rule or after we drag/drop to reorder) - to show that you can also wrap the entire form/component in a `<a-spin :spinning="loading"...`
- Change the add button icon from `save` to `plus-icon` or `plus` - it's because throughout the UI, the `plus` button is used to imply we're trying to add/create something
- Change 'Select Rule' placeholder to `Select API` and `Optional Description` to `Description`
- Minor css issue (not very important): the scrollbar does not scroll when I drag the items around (this is the same behaviour in the old UI, I was wondering if there is a way to improve this)

I'll test and review thoroughly tomorrow, did a quick pull and test. Except for the slow api list rendering, all look good.

File: src/views/iam/RolePermissionTab.vue (hunk 1 above)
Comment (RitchieVincent) [reply src/views/iam/RolePermissionTab.vue]: This has been completed as you requested now. However it's still a little slow, but I'm not sure what else can be done about it. It's needing to load 500+ list items, so it's understandable that it'll be slightly slow.

**Comment by RitchieVincent** [comment]
> @RitchieVincent well done, looks good. Few minor issues:
> 
> * When update API is called, show the loading/spinner (either on editing the deny/accept dropdown for a rule or after we drag/drop to reorder) - to show that you can also wrap the entire form/component in a `<a-spin :spinning="loading"...`
> * Change the add button icon from `save` to `plus-icon` or `plus` - it's because throughout the UI, the `plus` button is used to imply we're trying to add/create something
> * Change 'Select Rule' placeholder to `Select API` and `Optional Description` to `Description`
> * Minor css issue (not very important): the scrollbar does not scroll when I drag the items around (this is the same behaviour in the old UI, I was wondering if there is a way to improve this)
> 
> I'll test and review thoroughly tomorrow, did a quick pull and test. Except for the slow api list rendering, all look good.

What do you mean by the first point? I'm seeing the loading spinner on each edit of data.
I've now changed the icon and placeholder texts.
For the scroll issue, I don't think it's a CSS issue. It looks like something with the vue-draggable plugin. It'll need more investigation but that could probably come at a later date.

**Comment by yadvr** [comment]
I'll discuss tomorrow @RitchieVincent - I'll do a quick review/test and merge this

**Comment by yadvr** [comment]
Upstream PR sent now: https://github.com/apache/cloudstack-primate/pull/55

**Pull request closed without merging**